coord/src/rmr/rq.c

```diff
@@ -27,21 +28,71 @@ typedef struct MRWorkQueue {
   uv_async_t async;
 } MRWorkQueue;
 
+uv_thread_t loop_th;
+static char loop_th_running = false;
+static char loop_th_ready = false;
+uv_timer_t topologyTimer;
+uv_async_t topologyAsync;
+struct queueItem *pendingTopo = NULL;
+static inline struct queueItem *exchangePendingTopo(struct queueItem *newTopo) {
+  return __atomic_exchange_n(&pendingTopo, newTopo, __ATOMIC_SEQ_CST);
+}
+
+extern RedisModuleCtx *RSDummyContext;
+
+static void topologyTimerCB(uv_timer_t *timer) {
+  uv_async_send(&topologyAsync);
+}
+static void topologyAsyncCB(uv_async_t *async) {
+  struct queueItem *topo = exchangePendingTopo(NULL); // take the topology
+  if (!topo && MR_CheckTopologyConnections(true) == REDIS_OK) {
+    // We are connected to all master nodes. We can mark the event loop thread as ready
+    loop_th_ready = true;
+    RedisModule_Log(RSDummyContext, "verbose", "All nodes connected");
+  } else {
+    if (topo) {
+      // Apply new topology
+      RedisModule_Log(RSDummyContext, "verbose", "Applying new topology");
+      topo->cb(topo->privdata);
+      rm_free(topo);
+    }
+    // Finish this round of topology checks to give the topology connections a chance to connect.
+    // Schedule connectivity check in 1ms
+    RedisModule_Log(RSDummyContext, "verbose", "Waiting for all nodes to connect");
+    uv_timer_start(&topologyTimer, topologyTimerCB, 1, 0);
```

Comment:
Is this allowed to be called multiple times? I believe this is what currently happened right?

Reply:
`uv_timer_start` is not thread-safe, but a timer can be started multiple times. If the timer is already set this will reset it